### ISSUE TITLE
Support Audio File Extension Tags

### ISSUE THREAD

**Issue by jmarbach** [googleapis/google-cloud-ruby#2800]
Hi,

I am attempting to programmatically convert my audio files with ffmpeg before I create a Google Speech transcribe operation, however, I receive the following error that stops the execution of my program:

```
/.rvm/gems/ruby-2.5.1/gems/google-gax-1.5.0/lib/google/gax/api_callable.rb:369:in `rescue in block in retryable': GaxError Exception occurred in retry method that was not classified as transient, caused by 3:Request contains an invalid argument. (Google::Gax::RetryError)
```

I've debugged this issue extensively and I noticed that when I convert my audio files manually at the command line, rather than through a system call in my program, they produce a file without the application extension tag, e.g.

`audio/flac` vs `audio/x-flac`
`audio/wav` vs `audio/x-wav`
`audio/ogg` vs `application/octet-stream`
etc

The Google API is currently rejecting these perfectly fine audio files, resulting in a more tedious manual process me before I upload dozens of new transcription jobs.

Please support audio files with the extension tag; it will enable a lot more consumption!

**Comment by quartzmo** [comment]
Hi @jmarbach,

Thanks for opening this issue!

> Please support audio files with the extension tag; it will enable a lot more consumption!

Do you mean files **with** the extension tag or audio files **without** the extension tag?

**Comment by quartzmo** [comment]
I'm having a little trouble following exactly you are doing, can you provide more detail in the form of example code, audio file name/headers, and full backtrace? Most likely this issue can't be addressed in this client, but we're more than happy to take a look at it with you. (However, if a change in the Speech API service is what is needed, it might help to open a question on Stack Overflow with the [google-cloud-speech](https://stackoverflow.com/questions/tagged/google-cloud-speech) if you haven't already. Please link it here for future reference.)

**Comment by jmarbach** [comment]
Hey @quartzmo,

Thanks so much for offering to take a look! To answer your first question, I am hoping that the Speech-to-Text API will support files **with** the extension tag, e.g. `audio/x-flac`.

I ran into this problem because I need to convert my `MP3` encoded files to one of the following Speech-to-Text API compatible encodings:

`FLAC`, `LINEAR16`, `MULAW`, `AMR`, `AMR_WB`, `OGG_OPUS`, `SPEEX_WITH_HEADER_BYTE`


To automatically convert my `MP3` audio file, I wrote the following example code:

```
require "google_drive"
require "google/cloud/speech"
require "google/cloud/storage"

begin
  # Initialize Speech client
  speech = Google::Cloud::Speech.new(version: :v1p1beta1)

  # Initialize Storage client
  storage = Google::Cloud::Storage.new(
    project_id: "your-project-id",
    credentials: "your-service-account-credentials.json"
  )

  # Convert .mp3 file to .opus
  original_file_path = "/your-example-mp3-file-path.mp3"
  new_opus_path = original_file_path.gsub(".mp3", ".opus")

  # Context on the ffmpeg parameters: https://stackoverflow.com/a/49142056/1733600
  system("ffmpeg -i #{original_file_path} -c:a libopus -b:a 48k -vbr on -compression_level 10 -frame_duration 60 -application voip #{new_opus_path}")

  # Upload new file to Google Cloud Storage
  bucket = storage.bucket "example-bucket"
  new_file = bucket.create_file new_opus_path

  puts ""
  puts "Content Type:"
  puts new_file.content_type
  puts ""

  # Set operation config
  audio = { uri: new_file.to_gs_url }

  config = { encoding:          :OGG_OPUS,
             sample_rate_hertz: 48000,
             language_code:     "en-US",
             enable_speaker_diarization: false }

  # Start operation
  operation = speech.long_running_recognize config, audio
rescue => exception
  puts exception.backtrace
  raise
end
```

The script above produces the following...

```
Content Type:
application/octet-stream
```

And Error:

```
/Users/john/.rvm/gems/ruby-2.5.1/gems/google-gax-1.5.0/lib/google/gax/api_callable.rb:369:in `rescue in block in retryable'
/Users/john/.rvm/gems/ruby-2.5.1/gems/google-gax-1.5.0/lib/google/gax/api_callable.rb:358:in `block in retryable'
/Users/john/.rvm/gems/ruby-2.5.1/gems/google-gax-1.5.0/lib/google/gax/api_callable.rb:230:in `block in create_api_call'
/Users/john/.rvm/gems/ruby-2.5.1/gems/google-gax-1.5.0/lib/google/gax/api_callable.rb:260:in `block in create_api_call'
/Users/john/.rvm/gems/ruby-2.5.1/gems/google-cloud-speech-0.32.0/lib/google/cloud/speech/v1p1beta1/speech_client.rb:318:in `long_running_recognize'
script/support.rb:34:in `<main>'
Traceback (most recent call last):
	5: from script/support.rb:34:in `<main>'
	4: from /Users/john/.rvm/gems/ruby-2.5.1/gems/google-cloud-speech-0.32.0/lib/google/cloud/speech/v1p1beta1/speech_client.rb:318:in `long_running_recognize'
	3: from /Users/john/.rvm/gems/ruby-2.5.1/gems/google-gax-1.5.0/lib/google/gax/api_callable.rb:260:in `block in create_api_call'
	2: from /Users/john/.rvm/gems/ruby-2.5.1/gems/google-gax-1.5.0/lib/google/gax/api_callable.rb:230:in `block in create_api_call'
	1: from /Users/john/.rvm/gems/ruby-2.5.1/gems/google-gax-1.5.0/lib/google/gax/api_callable.rb:358:in `block in retryable'
/Users/john/.rvm/gems/ruby-2.5.1/gems/google-gax-1.5.0/lib/google/gax/api_callable.rb:369:in `rescue in block in retryable': GaxError Exception occurred in retry method that was not classified as transient, caused by 3:Request contains an invalid argument. (Google::Gax::RetryError)
```

By comparison, when I complete the encoding conversion at the command line by running:
```
ffmpeg -i "/original-file.mp3" -c:a libopus -b:a 48k -vbr on -compression_level 10 -frame_duration 60 -application voip "/new-file.opus"
```
The new .opus file Content Type is:
```
audio/ogg
```
And then when I upload this new file to Google Storage and start a new transcribe operation, I do not get any errrors.

Does this make sense? Do you still think it would be helpful to cross-post on Stack Overflow?

**Comment by blowmage** [comment]
It looks to me like the problem you are having is uploading a file with the correct content type. The default content type for an unknown binary file is `application/octet-stream`. So this makes me think the uploaded files are not using a well known file extension.

The Ruby Google Cloud Storage gem attempts to identify a default content type for a well known file extension when a storage file is created and is not provided a specific `content_type` value. We can see how `google-cloud-storage` will behave using the `mime-types` gem with the following code:

```ruby
require "mime/types"

MIME::Types.of "/new-file.opus" #=> []
```

As you can see, the `.opus` file extension is not well know. However, the `.ogg` file extension is well known:


```ruby
require "mime/types"

MIME::Types.of "/new-file.ogg" #=> [#<MIME::Type: audio/ogg>, #<MIME::Type: video/ogg>]
```

So, by default the `.opus` files are going to be created with the `application/octet-stream` content type, while `.ogg` files are going to be created with the `audio/ogg` content type. However, you can set any content type to any file, regardless of file extension, when creating a new GCS file:

```ruby
# Upload new file to Google Cloud Storage
bucket = storage.bucket "example-bucket"
new_file = bucket.create_file new_opus_path, content_type: "audio/ogg"
```

Also, be sure to look at the `details`, `status_details`, and `metadata` attributes on the error object for [more details](http://googleapis.github.io/google-cloud-ruby/docs/google-cloud-core/latest/Google/Cloud/Error.html) about why the error occurred.

**Comment by jmarbach** [comment]
@blowmage Thanks so much for providing all your help! You were right regarding the problem being related to the file that I am uploading. To solve this issue, I did the following:

1. I started specifying the `content_type` attribute to ensure that the file I was uploading matches a supported content type.
2. I started specifying an explicit destination path on file upload. The previous destination path had an extra `/` that was causing the Speech to not be able to recognize the new GCS file URL.

Feel free to close this issue.
